cloud/ec2_price.go

```diff
@@ -93,85 +94,154 @@ func getZone(ctx context.Context, client AWSClient, h *host.Host) (string, error
+	p, err := cpf.parseAWSPricing(out)
+	if err != nil {
+		return 0, errors.Wrap(err, "problem parsing aws pricing")
+	}
+	cpf.Lock()
+	defer cpf.Unlock()
+	cpf.ec2Prices[odInfo{os: osPriceName, instance: instance, region: region}] = p
+	return p, nil
+}
 
-	_, err := util.Retry(func() (bool, error) {
-		resp, err := client.Get(endpoint)
-		if resp != nil {
-			defer resp.Body.Close()
-		}
-		if err != nil {
-			return true, errors.Wrapf(err, "fetching %v", endpoint)
-		}
-		grip.Debug("Parsing on-demand pricing")
+func (cpf *cachingPriceFetcher) makeGetProductsInput(info odInfo) *pricing.GetProductsInput {
+	const match = "TERM_MATCH"
+	constructGetProductsInput := &pricing.GetProductsInput{
+		Filters: []*pricing.Filter{
+			{
+				Field: aws.String("ServiceCode"),
+				Type:  aws.String(match),
+				Value: aws.String("AmazonEC2"),
+			},
+			{
+				Field: aws.String("productFamily"),
+				Type:  aws.String(match),
+				Value: aws.String("Compute Instance"),
+			},
+			{
+				Field: aws.String("preInstalledSw"),
+				Type:  aws.String(match),
+				Value: aws.String("NA"),
+			},
+			{
+				Field: aws.String("tenancy"),
+				Type:  aws.String(match),
+				Value: aws.String("Shared"),
+			},
+			{
+				Field: aws.String("capacityStatus"),
+				Type:  aws.String(match),
+				Value: aws.String("UnusedCapacityReservation"),
+			},
+			{
+				Field: aws.String("instanceType"),
+				Type:  aws.String(match),
+				Value: aws.String(info.instance),
+			},
+			{
+				Field: aws.String("operatingSystem"),
+				Type:  aws.String(match),
+				Value: aws.String(info.os),
+			},
+			{
+				Field: aws.String("location"),
+				Type:  aws.String(match),
+				Value: aws.String(info.region),
+			},
+		},
+		ServiceCode: aws.String("AmazonEC2"),
+	}
+	return constructGetProductsInput
+}
 
-		if err = json.NewDecoder(resp.Body).Decode(&details); err != nil {
-			return true, errors.Wrap(err, "parsing response body")
+// Parse a JSON document like this one
+// {
+// 	"version": "20181031070014",
+// 	"terms": {
+// 		"OnDemand": {
+// 			"XCH2WN4F4MYH63N8.JRTCKXETXF": {
+// 				"termAttributes": {
+// 				},
+// 				"sku": "XCH2WN4F4MYH63N8",
+// 				"priceDimensions": {
+// 					"XCH2WN4F4MYH63N8.JRTCKXETXF.6YS6EN2CT7": {
+// 						"unit": "Hrs",
+// 						"rateCode": "XCH2WN4F4MYH63N8.JRTCKXETXF.6YS6EN2CT7",
+// 						"pricePerUnit": {
+// 							"USD": "0.8400000000"
+// 						},
+// 						"endRange": "Inf",
+// 						"description": "$0.840 per Unused Reservation Linux c3.4xlarge Instance Hour",
+// 						"beginRange": "0",
+// 						"appliesTo": []
+// 					}
+// 				},
+// 				"offerTermCode": "JRTCKXETXF",
+// 				"effectiveDate": "2018-10-01T00:00:00Z"
+// 			}
+// 		}
+// 	},
+// [...]
+func (cpf *cachingPriceFetcher) parseAWSPricing(out *pricing.GetProductsOutput) (float64, error) {
+	if len(out.PriceList) != 1 {
+		return 0, errors.Errorf("problem parsing price list %v", out.PriceList)
+	}
+	terms, ok := out.PriceList[0]["terms"].(map[string]interface{})
```

Comment:
also this is potentially ambigious... does the ok mean that terms is not nil, or that the map is in fact a map? probably ought to split this up.

Reply:
done